### PR TITLE
119 buttons

### PR DESCRIPTION
**Pull request by IssaHanou**
### Issue 
closes GH-119 and closes gh-167 (logging was merged in this)

## Contents
`button_events` in config, now generate buttons in the manage button section in front-end. these events are not shown in the puzzle overview.

The conditions are checked and if they resolve to true and the limit is not yet exceeded then the buttons is clickable, otherwise it is disabled. When clicked, these events and their actions are executed. 

Buttons are ordered alphabetically in front-end

## TODO
example with start/stop/pause  - does that work correctly

### Related issue
Related to GH-175 for device checking in conditions

### start, stop, pause
Currently implements start, stop, pauze, verder gaan.

Start is set to true when clicked. Only when start is clicked, can the rest be clicked. When pauze is clicked, its status is set to true. Only when pauze is true, can verder gaan be clicked. This sets the status of pauze and verder gaan back to false (it appears as a normal running game).

stop can be pressed any time after start, after which nothing else can happen anymore.

### CODE REVIEW

File: back-end/src/sciler/config/configHandler.go [hunk 1]
```diff
@@ -48,21 +48,17 @@ func generateDataStructures(readConfig ReadConfig) (WorkingConfig, []string) {
 	config.Puzzles = newPuzzles
 	newEvents, eventErrors := generateGeneralEvents(readConfig.GeneralEvents, &config)
 	config.GeneralEvents = newEvents
-	errorList = append(errorList, append(puzzleErrors, eventErrors...)...)
+	newButtonEvents, buttonEventErrors := generateButtonEvents(readConfig.ButtonEvents, &config)
+	config.ButtonEvents = newButtonEvents
+	errorList = append(errorList, append(append(puzzleErrors, eventErrors...), buttonEventErrors...)...)
```
Comment (GwennanS) [inline back-end/src/sciler/config/configHandler.go]: can you make it either append(append(append(puzzleErrors, eventErrors...), errorList...) , buttonEventErrors...)
or append(errorList, append(buttonEventErrors, append(puzzleErrors, eventErrors...)...)...)

File: back-end/src/sciler/handler/handler.go [hunk 2]
```diff
@@ -50,6 +50,7 @@ func (handler *Handler) msgMapper(raw Message) {
 		{
 			handler.updateStatus(raw)
 			handler.sendStatus(raw.DeviceID)
+			handler.sendFrontEndStatus(raw)
```
Comment (GwennanS) [inline back-end/src/sciler/handler/handler.go]: can sendStatus, sendFrontEndStatus and sendEventStatus be combined to sendStatusToFrontEnd

File: back-end/src/sciler/handler/handler_helper.go [hunk 3]
```diff
@@ -206,6 +250,22 @@ func (handler *Handler) getCameras() []map[string]string {
 	return cameras
 }
 
+// getButtons returns a list with button names
+func (handler *Handler) getButtons() []map[string]interface{} {
+	var buttons []map[string]interface{}
+	for _, btn := range handler.Config.ButtonEvents {
+		rule, ok := handler.Config.RuleMap[btn.ID]
+		if !ok {
+			logrus.Errorf("rule with id %s not found in rule map for button %s", rule.ID, btn.ID)
```
Comment (GwennanS) [inline back-end/src/sciler/handler/handler_helper.go]: is it possible to get to this error, and if so then it should have been checked reading the config in

File: back-end/src/sciler/handler/handler.go (hunk 2 above)
Comment (IssaHanou) [reply back-end/src/sciler/handler/handler.go]: no we want to split out the functionality

**Comment by mrlomar** [comment]
If it is not too much work, it would be nice if the order of the buttons would be the same as they are ordered in the configuration file.

**Comment by IssaHanou** [comment]
Currently, puaze and verder gaan have limit 1. then it works.

With setting limit to 0, and expecting one to alternate the other, you need to reset the status of pause to false when clicking verder gaan. this causes trouble as this requires the sending of instruction to change status to front-end -> which calls the send status, sending its new status back to back-end -> which acts on new status and executes pause button again. This is an issue, related to GH-137

**Comment by IssaHanou** [comment]
Currently implements `start`, `stop`, `pauze`, `verder gaan`. 

`Start` is set to true when clicked. Only when `start` is clicked, can the rest be clicked. When `pauze` is clicked, its status is set to true. Only when `pauze` is true, can `verder gaan` be clicked. This sets the status of `pauze` and `verder gaan` back to false (it appears as a normal running game). 

`stop` can be pressed any time after `start`, after which nothing else can happen anymore.

**Comment by mrlomar** [comment]
The information about when the escape room ends (done at xx:xx) is not showing when starting an escape room

File: back-end/resources/manuals/room_manual.md [hunk 4]
```diff
@@ -53,13 +61,103 @@ General events have the following properties:
...
+The conditions must include the value of the button on which it should execute. 
+The can also include extra parameters that the executing of the rule depends on, 
+like stop can only be pressed when start is already pressed.
+
+The buttons to manage the game state are configured in this sections.
+The front-end has output component `gameState` which keeps track of state of the game.
+The button event conditions can depend on the `gameState` and the actions should alter the `gameState`.
+
+The `gamestate` can have several states, which can be defined through the config. 
+Important states are: `opgestart`, `in spel`, `gepauzeerd` and `gestopt`, but more states can be used.
```
Comment (mrlomar) [inline back-end/resources/manuals/room_manual.md]: ~Importants states~ Example states

File: back-end/resources/manuals/message_manual.md [hunk 5]
```diff
@@ -96,7 +96,7 @@ are specific defined depending on the sender and receiver.
             - `instruction`
             - `instructed_by`
     - If type is `instruction`, then the then the message contents have
-             - `instruction` with value `reset` or `status update` or `test`
+             - `instruction` with value `reset` or `status update` or `test` or `setState`
```
Comment (mrlomar) [inline back-end/resources/manuals/message_manual.md]: `setState` should  be `set state` since we don't use camelcase for instructions

File: back-end/src/sciler/config/workingConfigTypes.go [hunk 6]
```diff
@@ -162,16 +163,16 @@ type InstructionSender interface {
...
 func (r *Rule) Finished() bool {
-	return r.Executed == r.Limit
+	return r.Executed == r.Limit && r.Limit != 0
 }
 
 // Execute performs all actions of a rule
 func (r *Rule) Execute(handler InstructionSender) {
 	for _, action := range r.Actions {
-		go action.Execute(handler)
+		action.Execute(handler)
```
Comment (mrlomar) [inline back-end/src/sciler/config/workingConfigTypes.go]: Why is this changed?

File: back-end/src/sciler/handler/handler_helper.go [hunk 7]
```diff
@@ -157,15 +168,46 @@ func (handler *Handler) sendEventStatus() {
...
 	handler.Communicator.Publish("front-end", string(jsonMessage), 3)
 }
 
-// getEventStatus returns json list with json objects with keys ["id", "status"]
+// getEventStatus returns a json list with json objects with keys ["id", "status"]
 // status is json object with key ruleName and value true (if executed == limit) or false
 func (handler *Handler) getEventStatus() []map[string]interface{} {
 	var list []map[string]interface{}
-	for _, rule := range handler.Config.RuleMap {
+	for _, rule := range handler.getPuzzleRules() {
```
Comment (mrlomar) [inline back-end/src/sciler/handler/handler_helper.go]: Why not use a cached map?

File: back-end/src/sciler/handler/handler_helper.go [hunk 8]
```diff
@@ -183,16 +238,16 @@ func (handler *Handler) getHints() map[string][]string {
 	return hints
 }
 
-// getEventDescriptions returns map of hints with puzzle name as key and list of hints for that puzzle as value
+// getEventDescriptions returns a map of hints with puzzle name as key and list of hints for that puzzle as value
 func (handler *Handler) getEventDescriptions() map[string]string {
 	events := make(map[string]string)
-	for _, rule := range handler.Config.RuleMap {
+	for _, rule := range handler.getPuzzleRules() {
```
Comment (mrlomar) [inline back-end/src/sciler/handler/handler_helper.go]: Why not use a cached map?

File: front-end/src/app/components/timer/timer.component.ts [hunk 9]
```diff
@@ -27,7 +27,7 @@ export class TimerComponent implements OnInit {
     const device = this.app.deviceList.getDevice("front-end");
     if (device !== null) {
       const status = device.status;
-      if (status.get("start") > 0 && status.get("stop") === 0) {
+      if (status.get("gameState") == "in spel") {
```
Comment (mrlomar) [inline front-end/src/app/components/timer/timer.component.ts]: If you hard code this state, point out in the manual that this state is required, or make this state configurable, or perhaps even better, do not check what the status is but what the state of the general timer is.

File: back-end/src/sciler/handler/handler_helper.go (hunk 8 above)
Comment (IssaHanou) [reply back-end/src/sciler/handler/handler_helper.go]: because the rulemap includes button events which should not be used here
